### PR TITLE
Hermite basis

### PR DESCRIPTION
**Pull request by AudreyBu**
## Description

Hermite basis which reconstruct a cubic polynomial using the values and derivatives at the endpoints of a given interval. Added a few lines in Affine.py to feed in the physical interval in the basis.fit function in order to account for the Jacobian $1/(b-a)$ when computing coefficients.

Reopened this new pull request to fix rebase issue and errors from previous commits.

## Linked Issue
Fixes # 14

## Implementation Checklist
- [x] My class inherits from `PolynomialBasis` (if applicable).
- [x] I have implemented all abstract methods (`evaluate_basis`, `fit`, etc.).
- [x] I have used `numpy` vectorization to avoid inefficient Python loops.
- [x] Documentation strings (docstrings) are provided for all public methods.

### 1. Mathematical Integrity
- [ ] **Convergence:** Does the implementation maintain expected accuracy? (Checked for Runge's phenomenon or stability).
- [ ] **Edge Cases:** Does it handle single points or non-equally spaced nodes correctly?
- [ ] **Domain Mapping:** Does the class correctly map the input domain to the basis domain (e.g., [-1, 1] for Chebyshev)?

### 2. API & Style
- [ ] **Method Signatures:** Methods match the Abstract Base Class exactly.
- [ ] **Naming:** Follows `CamelCase` for classes and `snake_case` for functions/variables.
- [ ] **Clean Code:** No "magic numbers"; constants are named and explained.

### 3. Testing Quality
- [ ] **Coverage:** Unit tests cover typical usage and edge cases.
- [ ] **Precision:** Tests use `np.testing.assert_allclose` instead of `==`.
- [ ] **Validation:** There is a test proving the basis can exactly reconstruct a polynomial of its own degree.

## Screenshots / Plots

### CODE REVIEW

File: src/numanalysislib/basis/hermite.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        x_nodes = np.asarray(x_nodes, dtype=float)
+        y_nodes = np.asarray(y_nodes, dtype=float)
+
+        if x_nodes.shape[0] != 2:
+            raise ValueError("HermiteBasis.fit expects the interval as [a, b].")
+        if y_nodes.shape[0] != self.n_dofs:
+            raise ValueError(f"HermiteBasis.fit expects {self.n_dofs} data entries [v_a, d_a, v_b, d_b].")
+
+        # Determine scaling for derivative coefficients. 
+        # If a physical interval is provided (from an Affine wrapper), use that; otherwise infer from x_nodes.
```
Comment (viktorya-grigoryan) [inline src/numanalysislib/basis/hermite.py]: What if we provide only the physical interval? Looks like we would need to give x_nodes too anyway, even though we don't use it in that case

File: src/numanalysislib/basis/hermite.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
+"""
+Hermite cubic polynomial basis functions.
+
+p(x) = v_a * h_0(x) + d_a * h_1(x) * (b-a) + v_b * h_2(x) + d_b * h_3(x) * (b-a)
+"""
```
Comment (mtezzele) [inline src/numanalysislib/basis/hermite.py:5]: Move this in line 13 as the description of the class. At the module level (here) you can just write a brief description.

**Comment by mtezzele** [comment]
@max-collins please have a look at the edit in the Affine class

File: src/numanalysislib/basis/hermite.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        x3 = x2 * x
+
+        if index == 0:
+            return 2.0 * x3 - 3.0 * x2 + 1.0
+        elif index == 1:
+            return x3 - 2.0 * x2 + x
+        elif index == 2:
+            return -2.0 * x3 + 3.0 * x2
+        elif index == 3:
+            return x3 - x2
```
Comment (mtezzele) [inline src/numanalysislib/basis/hermite.py]: use the python `match-case` statement instead of `if-elif`

**Comment by max-collins** [comment]
> @max-collins please have a look at the edit in the Affine class

Fix looks good to me.
